### ISSUE TITLE
Option to overwrite existing overlapping output with Gradle build cache

### ISSUE THREAD

**Issue by gluxon** [gradle/gradle#20467]
### Context

Our organization uses Gradle for frontend TypeScript builds and has found it to be as great as advertised. We're using Gradle to deliver "_elegant, blazing-fast builds_" on our Continuous Integration (CI) pipelines.

One area we're looking to improve is remote build cache usage during local development.

At the moment there's 2 ways developers can run TypeScript locally in our project:

1. Through the `compileTypescript` task in Gradle. We've configured this to retrieve `.d.ts` definitions from a remote build cache. If the build cache can't be used, it falls back to a `tsc` (TypeScript compiler) `Exec` task.
2. Through TypeScript's builtin watcher. This is typically invoked on the command line through `npm run compile` or  `tsc --watch`.

### Current Behavior
If a developer runs `tsc --watch`, this unfortunately actually causes `./gradlew compileTypescript` to run much slower afterwards. Gradle sees output created earlier from `tsc --watch` and will not use the remote build cache due to overlapping outputs detection.

![Screen Shot 2022-04-13 at 12 20 45 PM](https://user-images.githubusercontent.com/906558/163232246-f8f238f4-7b85-4727-9972-f249fa667055.png)

Copying the message in the screenshot for easier reading:

> Overlapping outputs: Gradle does not know how file 'lib/types/index.js.map' was created (output property '...'). Task output caching requires exclusive access to output paths to guarantee correctness (i.e. multiple tasks are not allowed to produce output in the same location).

### Expected Behavior

We'd find it useful if some tasks could specify overlapping output from non-Gradle processes as acceptable for that task. I'm imagining an `outputs.dangerouslyOverwriteExistingOutputWithBuildCache` option that relaxes the check above.

We've confident that our per-project `compileTypescript` task has exclusive access to the `lib` dir of that project. Even if there's other files in `lib`, in every case we'd want the `compileTypeScript` task to replace `lib` with the outputs of a previous remote build cache entry. (Assuming a cache it.)

### Alternatives

We've explored `./gradlew compileTypescript --continuous`, but observed it to be much slower than `tsc --watch` when picking up changes. The `tsc --watch` process is able to perform incremental compilation in-memory in a way the `:compileTypescript` exec task is not.

---

I'd be happy to help implement this new option if we think it matches Gradle's design philosophy.

**Comment by stigkj** [comment]
Couldn't you have `tsc --watch` use some other output directory?

**Comment by gluxon** [comment]
That's a good question. Unfortunately not without significant caveats to performance and usability. The main blockers are:

## 1. Other Tooling

**The TypeScript output directory isn't looked at exclusively by TypeScript.** To signal that an npm package has typings, the `package.json` `types` field is set. 

```jsonc
// packages/example/package.json
{
  "name": "example",
  "main": "lib/index.js",
  "types": "lib"
}
```

To maintain correctness at local development time, this field needs to be edited to the new directory after running `tsc --watch`, and edited back after running `./gradlew compileTypescript`.

```jsonc
// packages/example/package.json
{
  "name": "example",
  "main": "lib/index.js",
  "types": "lib-for-tsc-watcher"
}
```

ESLint and VS Code also read `.d.ts` files, and they would otherwise go out of date (producing falsey errors) if you don't edit the `package.json` field. However, editing `package.json` has its own problems:

- Since `package.json` is committed to source, this would cause hundreds of files to appear in `git status`. I think this would be a significant usability regression for our developers. They'd have to tiptoe around not committing these modifications.
- The `package.json` file is an input to most other Gradle build tasks we've set up. This would cause build cache misses for other tasks.

## 2. Project References

TypeScript projects in a monorepo can form a directed acyclic graph of dependencies. The TypeScript team refers to this feature as [project references](https://www.typescriptlang.org/docs/handbook/project-references.html). To change the output directory, the [`compilerOptions.outDir`](https://www.typescriptlang.org/tsconfig#outDir) field of every `tsconfig.json` file in every project would have to be changed. There isn't a command line argument to convince the TypeScript watcher to do this for multi-project builds.

We considered mirroring the `packages/<name>/tsconfig.json` files at `packages/<name>/tsconfig.watcher.json` with the only change being a different `outDir`. A new build tool/step needs to verify `packages/<name>/tsconfig.json` and `packages/<name>/tsconfig.watcher.json` don't go out of sync. This alone would be tractable, but there's still the `package.json` problem mentioned above, and the incremental builds problem mentioned below.

## 3. Incremental TypeScript Builds

Suppose we did create `tsconfig.watcher.json` files, and a script is written that performs the following:

1. Runs `./gradlew compileTypescript` to populate all build caches.
2. Copies all `packages/<name>/lib` directories into `packages/<name>/lib-for-tsc-watcher`.
3. Runs `tsc --watch` with the `tsconfig.watcher.json` files mentioned above in play.

It turns out `tsc --watch` will still perform a full recompilation of the entire dependency graph, avoiding anything populated by Gradle build caches. This is because there's also a  [`packages/<name>/tsconfig.tsbuildinfo` file](https://www.typescriptlang.org/tsconfig#tsBuildInfoFile) at play. This file contains a copy of the compiler options `tsc` last ran with. By changing `compilerOptions.outDir`, the compiler options have changed and a full rebuild is necessary. The `tsconfig.tsbuildinfo` generated from the `./gradlew compileTypescript` task was for `tsconfig.json`, not `tsconfig.watcher.json`.

Suppose we perform a find/replace for `s/lib/lib-for-tsc-watcher/g` in `tsconfig.tsbuildinfo` as well. That would probably work, but we'd then rely on the fact that TypeScript doesn't change the format of `tsconfig.tsbuildinfo` in the future. This is an internal file without a defined structure. In theory, I could see `tsc` adding a hash of the `compilerOptions` object in the future for validation, which would invalidate any hackery done here.

### 4. Two-way Syncing

Even if we were to do all of the above successfully and the mitigate the significant developer experience regressions, running `./gradlew compileTypescript` would be slower than necessary unless the `lib-for-tsc-watcher` changes were synced back into `lib`. This requires running `s/lib-for-tsc-watcher/lib` on `tsconfig.tsbuildinfo` files again to revert.

We'd also need to determine whether or not `lib` or `lib-for-tsc-watcher` is the more up to date directory of the two.

**Comment by asodja** [comment]
In Gradle 7.5 we improved the continuous mode to be much more responsive, check [release notes](https://docs.gradle.org/7.5/release-notes.html#continuous-build). Could you maybe give it a try? I wonder if that solves the responsiveness issue compared to `tsc --watch`.

Regarding the `outputs.dangerouslyOverwriteExistingOutputWithBuildCache = true` flag: that could cause incorrectness in Gradle build and bring other problems, so we are not very keen to move in that direction.

**Comment by gluxon** [comment]
> In Gradle 7.5 we improved the continuous mode to be much more responsive, check [release notes](https://docs.gradle.org/7.5/release-notes.html#continuous-build). Could you maybe give it a try? I wonder if that solves the responsiveness issue compared to `tsc --watch`.

The continuous mode changes in Gradle 7.5 definitely look like improvements, but its performance is still much lower than `tsc --watch`.

## Task Graph Re-evaluation

One reason for the significant performance difference is because continuous mode re-evaluates the entire task graph regardless of what task's input files were changed.

To elaborate with an example, a common setup is an application subproject that depends on several library subprojects.

```mermaid
flowchart BT
  app --> libraryA
  app --> libraryB
  app --> libraryC
```

If any input files to the `app:compileTypescript` task are edited, continuous mode performs up-to-date checking on the tasks in `libraryA:compileTypescript`, `libraryB:compileTypescript`, and `libraryC:compileTypescript` as well. For large graphs with many libraries depending on each other, this can add 30s-60s of delay before the actual edited subproject's tasks run.

In comparison, the `tsc --watch` command is able to detect that only the `app` project was edited and only recompute that.

## In-Memory

Once up-to-date checking is finished, it takes another minute for `tsc` to finish executing on the `app` subproject. When the `tsc` binary loads, it has to read all the `.d.ts` files in all dependent projects each time. The `tsc --watch` CLI is already running and has this in-memory, so this additional one minute delay becomes zero.

I tested a minor white-space only change to an application in my project just now and the total time was `2m 16s`.

## Future

I do see a world where Gradle's continuous mode could be smarter about what task inputs were edited and only recompute that, but I'm not sure how continuous mode could be on-par with the in-memory performance of `tsc --watch`. Open to ideas there, but I think we'd need a change to `tsc --watch` that keeps everything in memory but waits for Gradle to tell it what project to recompute.

> Regarding the `outputs.dangerouslyOverwriteExistingOutputWithBuildCache = true` flag: that could cause incorrectness in Gradle build and bring other problems, so we are not very keen to move in that direction.

Could I ask a few questions to understand this better? I see how for some tasks this flag would be dangerous and cause incorrectness. In this scenario, what incorrectness in the build and other problems would this flag bring?

**Comment by gluxon** [comment]
I ran a quick test to see how [turbo](https://turbo.build/) works in this scenario.

Running the first build:

```
❯ turbo build
• Running build
• Remote caching disabled
build: cache miss, executing 2b75896cbed4d74d
build: 
build: > turbo-test@1.0.0 build /Users/gluxon/Developer/turbo-test
build: > tsc
build: 

 Tasks:    1 successful, 1 total
Cached:    0 cached, 1 total
  Time:    1.334s 
```

If I make an edit to one of the output files:
```
~/Developer/turbo-test main
❯ echo "console.log('test')" >> lib/index.js
```

The build does use the remote cache to reset that edit:

```
❯ turbo build
• Running build
• Remote caching disabled
build: cache hit, replaying output 2b75896cbed4d74d
build: 
build: > turbo-test@1.0.0 build /Users/gluxon/Developer/turbo-test
build: > tsc
build: 

 Tasks:    1 successful, 1 total
Cached:    1 cached, 1 total
  Time:    50ms >>> FULL TURBO
```

Our team is pretty committed to Gradle at this point, but I thought this would be an interesting data point to show.

**Comment by asodja** [comment]
Thanks for trying continuous mode. So from what you wrote, I understand that it's not an option for you.

One quote from your post: 
> For large graphs with many libraries depending on each other, this can add 30s-60s of delay before the actual edited subproject's tasks run.

Not sure if this is configuration time or checking outputs. If it's the later we would love a reproducer, so we can maybe try to optimize this. 
If it's configuration time, we are actively working on configuration cache and project isolation that will make that faster.

----

Regarding how Gradle handle incrementality:
It's important to know that both inputs and outputs will affect if Gradle will use a remote cache.

1. First issue is editing outputs:
Gradle will in case output files are modified report `Overlapping outputs`. But interestingly if you delete all created/modified files, Gradle will work the same way as turbo.
So to reproduce turbo behavior you could add an additional task:
```
tasks.register("cleanupTscTask") {
    doNotTrackState("Not worth caching")
    doLast {
        // if tsc --watch run, delete all modified/created files or simply delete outputs of compileTypescript
    }
}

tasks.named("compileTypescript) {
    dependsOn("cleanupTscTask")
}
```
You can even just use `./gradlew clean compileTypescript` and Gradle will happily get files from the remote cache the same way as turbo.
So technically we could probably change a behavior, and if we see modified outputs we could overwrite them. But this needs more consideration.

2. But there is a bigger issue, you probably also change inputs when using `tsc --watch`:
A cache key for getting outputs for compileTypescript are its inputs, imagine a function like:
```
function getFromRemoteCache(inputs) {
    key = calculateKey(inputs)
    return remoteCache.get(key)
}
```
So if you edit inputs of compileTypescript while running `tsc --watch` then compileTypescript won't be able to find remote outputs. I am sure turbo has a similar problem when inputs change, they might have more coarse-grained caching though.

---
So if I understand your expectation of flow is:
1. Compile ts with `./gradlew compileTypescript`
2. Run `tsc --watch` and update source files
3. Run `./gradlew compileTypescript` ->  everything is up to date

And 3. can't happen unless there is some integration between `tsc` and `Gradle`, not because outputs are overwritten, but because inputs are also changed and you would need somehow to tell Gradle that new inputs match new outputs.

One way you could fix that is by using `doNotTrackState()`, which basically means `compileTypescript` will always run and you would handle caching on your own inside a task (or you would call some other tool like turbo to handle that for you).

**Comment by gluxon** [comment]
> Thanks for trying continuous mode. So from what you wrote, I understand that it's not an option for you.

Thanks for making it! Continuous mode is indeed helpful for other tasks.

> Not sure if this is configuration time or checking outputs. If it's the later we would love a reproducer, so we can maybe try to optimize this. If it's configuration time, we are actively working on configuration cache and project isolation that will make that faster.

It takes about ~15 seconds in our projects when all subprojects and tasks need to be configured. I do think most of the time is coming from task input/output checking. There's some work we can do to improve our own configuration time, but I don't believe that change alone would improve continuous mode performance.

I do agree a smaller reproduction of the problem would be good to verify either way. I can spin that up. In the meantime, I grabbed some metrics from our build scans to try and get a better idea of the configuration time vs input/output checking breakdown.

Here's a build scan from our private Gradle Enterprise instance showing configuration time on `./gradlew cleanTypescript compileTypescript`.

<img width="679" alt="Screenshot 2023-02-03 at 6 02 39 PM" src="https://user-images.githubusercontent.com/906558/216727663-6ae1cf45-ef35-4c65-914a-076891c9c4b8.png">

Here's task input fingerprinting time. In total the command took `49.198s`, so fingerprinting likely benefited from parallelism. (The `gradleEnterprise.buildScan.captureTaskInputFiles` flag was not set in case that's relevant.)

<img width="835" alt="Screenshot 2023-02-03 at 6 03 30 PM" src="https://user-images.githubusercontent.com/906558/216727766-517c0356-2a4e-40bb-a670-f2eca03cb255.png">

> Regarding how Gradle handle incrementality: It's important to know that both inputs and outputs will affect if Gradle will use a remote cache.
> 
> 1. First issue is editing outputs:
>        Gradle will in case output files are modified report `Overlapping outputs`. But interestingly if you delete all created/modified files, Gradle will work the same way as turbo.
>        So to reproduce turbo behavior you could add an additional task:
> 
> 
> ```
> tasks.register("cleanupTscTask") {
>     doNotTrackState("Not worth caching")
>     doLast {
>         // if tsc --watch run, delete all modified/created files or simply delete outputs of compileTypescript
>     }
> }
> 
> tasks.named("compileTypescript) {
>     dependsOn("cleanupTscTask")
> }
> ```
> 
> You can even just use `./gradlew clean compileTypescript` and Gradle will happily get files from the remote cache the same way as turbo.

I've been recommending the developers on my team to run `./gradlew cleanTypescript compileTypescript` when they're on a known good commit before any modifications have been made to the project. The `cleanTypescript` task isn't as smart as the `cleanupTscTask` you've defined above though, since it isn't aware of whether or not `tsc --watch` has ran before. Our clean task simply clears out the outputs dir unconditionally.

The downside of always running `cleanupTypescript` or `cleanupTscTask` is that it opts out of [TypeScript's incremental compilation](https://www.typescriptlang.org/tsconfig#incremental). If the build doesn't hit the remote cache, developers end up recompiling everything `tsc --watch` may have already compiled before since the clean task evicts the `*.tsbuildinfo` files from disk.

Right now we have to train developers to:

- Run `./gradlew cleanTypescript compileTypescript` if `tsc --watch` was ran previously, and they're on a fresh commit without any changes.
- Run `./gradlew compileTypescript` if they've made changes that are unlikely to hit the cache. (Such as modifying a subproject far down the dependency graph.)

The goal I'm chasing is to have my teammates always be able to run `./gradlew compileTypescript` and get the best possible performance.

> So technically we could probably change a behavior, and if we see modified outputs we could overwrite them. But this needs more consideration.

> 
> 2. But there is a bigger issue, you probably also change inputs when using `tsc --watch`:
>     A cache key for getting outputs for compileTypescript are its inputs, imagine a function like:
> 
> ```
> function getFromRemoteCache(inputs) {
>     key = calculateKey(inputs)
>     return remoteCache.get(key)
> }
> ```
> 
> So if you edit inputs of compileTypescript while running `tsc --watch` then compileTypescript won't be able to find remote outputs. I am sure turbo has a similar problem when inputs change, they might have more coarse-grained caching though.

Yup, you're right. I agree it's usually the case that inputs will change as developers are running `tsc --watch`. I think it's expected that `compileTypescript` won't ever hit the remote cache in this case though. The edits in theory should be new and never seen before.

Once developers have `git push`'ed, merged their PR, and then restart their work after branching from the `main` branch, I'd expect `compileTypescript` hits the cache again.

> So if I understand your expectation of flow is:
> 
>     1. Compile ts with `./gradlew compileTypescript`
> 
>     2. Run `tsc --watch` and update source files
> 
>     3. Run `./gradlew compileTypescript` ->  everything is up to date since `tsc --watch` compiled it
> 

That's roughly right, but I might amend this a bit:

1. Compile ts with `./gradlew compileTypescript`
2. Run `tsc --watch` and update source files.
3. Run `git push` and open a PR and merge into `main`.

The next day:

1. Run `git switch main-cached && git pull`.
2. Run `./gradlew compileTypescript` <-- Expecting this to be fast if the commit on `main-cached` was built.

The `main-cached` branch is a special branch we have that intentionally lags behind `main`. It gets updated when the CI workflow passes on for that commit on `main`, so `main-cached` is guaranteed to be fully cached. I think this is a common pattern in many repositories.

At the moment the `./gradlew compileTypescript` command recompiles the whole repo (taking ~15-20 minutes) the next day since `tsc --watch` was ran on the previous day.

> But unfortunately 3. can't happen unless there is some integration between `tsc` and `Gradle`. Not because outputs are overwritten, but because inputs are also changed and you would need somehow tell Gradle that new inputs match new outputs.

There's no API for this at the moment, right? Adding a wrapper around `tsc --watch` that tells teaches Gradle about its outputs would be another good way to solve this problem. I'm not aware of any existing ways to get the ideal workflow without making significant tradeoffs elsewhere.

> One way you could fix that is by using `doNotTrackState()`, which basically means `compileTypescript` will always run and you would handle caching on your own inside a task

Reading the documentation for `doNotTrackState`, this would out out of the remote build cache too?

> or you would call some other tool from a task like turbo to handle that for you

We actually really like the Gradle and wouldn't want to delegate this to Turbo. 😅 One of the reasons is because the build scans make it very easy to diagnose cache misses, which would be harder to do in Turbo.

**Comment by gluxon** [comment]
Your examples made me realize we could get somewhat close to the ideal workflow with:

```groovy
tasks.register("moveTypeScriptOutputs") {
    doNotTrackState("Not worth caching")
    doLast {
        // if tsc --watch run, move tsconfig.tsbuildinfo to tsconfig.tsbuildinfo-temp and lib to lib-temp
    }
}
```

```groovy
tasks.named("compileTypescript") {
    dependsOn("moveTypeScriptOutputs")
    doLast {
        // if lib-temp exists, move to lib
        // if tsconfig.tsbuildinfo-temp exists, move to tsconfig.tsbuildinfo
        exec { executable "tsc" }
    }
}
```

Moving the outputs elsewhere allows Gradle to use the remote build cache. Moving the outputs back in allows `tsc` to take advantage of its own incremental compilation behavior.

This gets close, but still regresses some other part of the existing workflow. Namely, it makes `./gradlew compileTypescript` slower on tasks that were previously skipped as up-to-date. In this world, Gradle would always run `tsc` on all subprojects. Those `tsc` processes would exit realizing no work is necessary, but it would still take time to fork into `tsc` ~800 times for each subproject.

The existing setup skips `tsc` invocations entirely by seeing the inputs and outputs were up to date locally.

**Comment by gluxon** [comment]
A few months after the last comment on this issue, @lptr filed https://github.com/gradle/gradle/issues/26702. That would address the performance problems with continuous mode mentioned here. 🙂
